src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ClientService.java

```diff
@@ -493,13 +495,27 @@ public ClientType deleteTlsCertificate(ClientId id, String certificateHash)
      * @throws ClientNotFoundException
      * @throws CannotRegisterOwnerException
      * @throws ActionNotPossibleException
+     * @throws InvalidMemberClassException
+     * @throws InvalidInstanceIdentifierException
      */
     public void registerClient(ClientId clientId) throws GlobalConfOutdatedException, ClientNotFoundException,
-            CannotRegisterOwnerException, ActionNotPossibleException {
+            CannotRegisterOwnerException, ActionNotPossibleException, InvalidMemberClassException,
+            InvalidInstanceIdentifierException {
 
         auditDataHelper.put(clientId);
 
         ClientType client = getLocalClientOrThrowNotFound(clientId);
+
+        String instanceIdentifier = client.getIdentifier().getXRoadInstance();
+        if (!instanceIdentifier.equals(globalConfFacade.getInstanceIdentifier())) {
+            throw new InvalidInstanceIdentifierException(INVALID_INSTANCE_IDENTIFIER + instanceIdentifier);
+        }
+
+        String memberClass = client.getIdentifier().getMemberClass();
+        if (!isValidMemberClass(memberClass)) {
+            throw new InvalidMemberClassException(INVALID_MEMBER_CLASS + memberClass);
+        }
```

Comment:
There are three methods in _ManagementRequestSenderService_ that take in a _ClientId_ parameter
* sendClientRegisterRequest
* sendClientUnregisterRequest
* sendOwnerChangeRequest

Would it make sense to check the validity of the _memberClass_ in all of these cases? Maybe moving the whole checking logic into _ManagementRequestSenderService_ methods listed above.

Or another way would be to just add the check to _unregisterClient_ and _changeOwner_ methods within _ClientService_.

Reply:
Originally I didn't include the member class check in `sendClientUnregisterRequest` and `sendOwnerChangeRequest`, because both requests require that the client has already been registered on the Security Server. And since the `sendClientRegisterRequest` has the check, registering a client with invalid member class should not be possible. So instead of including the check in all the three management requests, it's only included in the request that is a precondition for the two other requests.